### DISCUSSION TITLE
Interviews: Depth-First vs Breadth-First

### DISCUSSION THREAD

**Discussion by Sk8b04rd3r** [community/community#70105] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Product Feedback

### Body

I went through a bootcamp and wasn't taught searching very well. Can someone explain Depth-First vs Breadth-First?

**Comment by AneeqMalik** [comment]
@Sk8b04rd3r bro bassically the concept is pretty straigthforward. Consider a Depth-first search as going directly to the leave nodes from the root before moving to the next branch of the tree consider it as a single way direction at a time then reverse then single. For the  Breadth-First it is pretty similar to Level-order traversal where you have to explore all the nodes on the same level before going to the next level.

**Comment by fredschuck** [comment, marked as answer]
![image](https://github.com/community/community/assets/10211457/4b89fbb1-33a1-42d6-bbd8-89e7b67a3a72)


## Depth-First
This is a traversal algorithm that explores as far as possible along a branch before backtracking. It starts at an initial node and explores deeper into the graph or tree, visiting all connected nodes along a path before backtracking.

One thing to remember is that depth-first can be implemented using recursion.

The idea is to go as deep as possible before returning to explore other branches. You want to use this, for example, when trying to find nodes that are connected to each other or if you want to find what leaves or nodes are furthest away from the root. 

```
depth_first_search( node )
   if node is null:
        return

    mark node as visited
     < you can print the node's value or do something with this node here if you want >

    for each neighbor in neighbors of node:
        if neighbor is not visited:
            depth_first_search( neighbor )
```


## Breadth-First
This traversal algorithm explores all neighbors of a node before moving to their child nodes. It starts at an initial node and explores all its neighbors first, then moves to the neighbors' neighbors, and so on, level by level.

On the other hand, breadth-first is implemented using a queue data structure.

Here, the idea is to explore nodes layer by layer, ensuring that all nodes at a particular level are visited before moving to the next level.
It's useful in finding the shortest path in unweighted graphs. One way to think of it is exploring a set of nodes by moving outward in all directions one step at a time, ensuring that you explore all nearby options before moving further away. You can use this when trying to find   all neighbors first or what the shortest path is. 

```
breadth_first_search( startNode ):
    create an empty queue
    create an empty set to track visited nodes
    add startNode to the queue
    mark startNode as visited

    while the queue is not empty:
        current = remove a node from the front of the queue
         < you can print the node's value or do something with this node here if you want >

        for each neighbor in neighbors of current:
            if neighbor is not visited:
                add neighbor to the queue
                mark neighbor as visited
```

**Reply by Sk8b04rd3r** [reply]
woah thank you

**Reply by fredschuck** [reply]
No problem.  The statements underneath the description is pseudocode to help you write in whatever language you are programming in.
